### PR TITLE
Add sparsity and premium to schools

### PR DESCRIPTION
**Pull request by ngw**
Add data to schools' table in analytics

### CODE REVIEW

**Comment by ethanmills** [comment]
Missing a migration, no?

**Review by ethanmills — DISMISSED** [review]

File: app/services/analytics/ecf_school_service.rb [hunk 1]
```diff
@@ -41,6 +44,8 @@ def update_school_analytics
                 delivery_partner: school.active_partnerships.first&.delivery_partner&.name || school.partnerships.first&.delivery_partner&.name,
                 chosen_cip: school.school_cohorts.find { |sc| sc.cohort_id == current_cohort_id }&.core_induction_programme&.name,
                 active_participants: school.school_cohorts.find { |sc| sc.cohort_id == current_cohort_id }&.active_ecf_participants&.any? || false,
+                sparsity: school.local_authority_district.district_sparsities.find { |ds| ds.start_year <= current_cohort_year && ds.end_year > current_cohort_year }&.any?,
```
Comment (ethanmills) [inline app/services/analytics/ecf_school_service.rb:47]:
```suggestion
                sparsity: school.local_authority_district&.district_sparsities.find { |ds| ds.start_year <= current_cohort_year && ds.end_year > current_cohort_year }&.any?,
```

File: app/services/analytics/ecf_school_service.rb [hunk 2]
```diff
@@ -7,13 +7,16 @@ def update_school_analytics
...
+        current_cohort_year = Cohort.current.start_year
 
         School.eligible_or_cip_only.includes(
           :nomination_emails,
           :induction_coordinators,
           :partnerships,
-          school_cohorts: :active_ecf_participants,
           active_partnerships: %i[lead_provider delivery_partner],
+          local_authority_district: :district_sparsities,
+          pupil_premiums: :participant_profiles,
```
Comment (ethanmills) [inline app/services/analytics/ecf_school_service.rb:18]: Why the participant profiles here?

File: app/services/analytics/ecf_school_service.rb [hunk 3]
```diff
@@ -41,6 +44,8 @@ def update_school_analytics
                 delivery_partner: school.active_partnerships.first&.delivery_partner&.name || school.partnerships.first&.delivery_partner&.name,
                 chosen_cip: school.school_cohorts.find { |sc| sc.cohort_id == current_cohort_id }&.core_induction_programme&.name,
                 active_participants: school.school_cohorts.find { |sc| sc.cohort_id == current_cohort_id }&.active_ecf_participants&.any? || false,
+                sparsity: school.local_authority_district&.district_sparsities&.find { |ds| ds.start_year <= current_cohort_year && ds.end_year > current_cohort_year }&.any? || false,
+                pupil_premium: school.pupil_premiums.find { |pp| pp.start_year == current_cohort_year }&.uplift?,
```
Comment (jamesrose) [inline app/services/analytics/ecf_school_service.rb:48]:
```suggestion
                sparsity: school.sparsity_uplift?(current_cohort_year),
                pupil_premium: school.pupil_premium_uplift?(current_cohort_year),
```

I'm pretty sure these are doing the same thing.

Comment (ngw) [reply app/services/analytics/ecf_school_service.rb:48]: yeah that was my first try. It's n+1 hell though

Comment (ethanmills) [reply app/services/analytics/ecf_school_service.rb:48]: Yeah, that would take us from 50 queries to 50k queries

**Pull request merged**